### ISSUE TITLE
Problem with SQL array parameters

### ISSUE THREAD

**Issue by PetrSilar** [jtv/libpqxx#728]
TL;DR: It is possible to use `vector` of `string`s as SQL array query parameter but it's not possible to use `vector` of `string_view`s (fails to compile) nor `zview`s (creates invalid parameter). 
&NewLine;
E.g.:
```c++
const auto query = "DELETE FROM some_table WHERE id = ANY($1)"s;
const auto id1 = "123-456"s;
const auto id2 = "789-000"s;
```
&NewLine;
Using `std::vector<std::string>` is ok:
```c++
std::vector<std::string> ids{id1, id2};
tr.exec_params0(query, ids);
```
&NewLine;
Using `std::vector<std::string_view>` will not compile because there is no `string_traits<std::string_view>::to_buf()`:
```c++
std::vector<std::string_view> ids{id1, id2};
tr.exec_params0(query, ids);
```
&NewLine;
Using `std::vector<pqxx::zview>` will compile but will fail at runtime because the parameter will be `{"` because `string_traits<zview>::to_buf()` returns pointer to the end of the buffer and container's `to_buf()` creates garbage:
```c++
std::vector<pqxx::zview> ids{id1, id2};
tr.exec_params0(query, ids);
```
&NewLine;
Using `std::vector<std::reference_wrapper<const std::string>>` will compile but will not link because of missing implementation of the traits:
```c++
std::vector<std::reference_wrapper<const std::string>> ids{id1, id2};
tr.exec_params0(query, ids);
```
&NewLine;
Please note, using scalar parameters of either string-like type (excl. `cref`) works well (because `string_traits<type>::into_buf()` is used instead).
&NewLine;
IMO vector of any string-like type should work since the value which will be finally passed to *libpq* from it is created inside of *libpqxx* (having its own buffer). Also it would be great if *libpqxx* accepts any range (actually `std::ranges::forward_range`) when there is c++20 in 8.x already instead of a `vector` only.

**Comment by PetrSilar** [comment]
Personally, I would copy everything into internal buffer which actually happens for all non-string types (numbers, UUIDs, timestamps, ...) anyway. At least on PCs, it is insignificant amount of memory and such copying cannot affect overall performance of requests processing. That could also simplify traits requirements, maybe there would be no need for `zview` at all. Memory of buffers for a *libpq* API call could be allocated from a pool so there would be the only memory allocation per request.
&NewLine;
Speaking about ranges, there is concept of forward range where you can keep iterators for later access and also concept of contiguous range guaranteeing (obviously :relaxed:) its data is in contiguous block of memory.
&NewLine;
Back to SQL array parameters. IMO no current limitation of the library is important much (or necessary). You have to iterate over all elements, convert them into escaped strings and finally create (within own internal buffer) a string `{"elem-one-string-repre","elem-two-string-repre","etc"}` which is eventually consumed by some *libpq* API call.
&NewLine;
Thank you for fast fix of `zview`'s traits! :thumbsup:

**Comment by jtv** [comment]
Thanks for the report!  The good news there is that it was _so_ broken that nobody could ever have relied on the functionality.

Now for the other stuff...  Yes I do copy data into a buffer where necessary, and I do allocate just a single buffer where possible.  The main part I dislike is that there's no feedback to the application developer about the costs, especially when those are not obvious.  As you say it's going to be small in proportion to the business of sending the data across the socket and perhaps the network, parsing it, and executing the SQL.  On the other hand, OS developers in the past have found cases where engineers at too many layers have reasoned like this, accumulating significant overhead that they could optimise away.  Hence all the "zero-copy" efforts.

**Comment by PetrSilar** [comment]
Yes, I understand you very well. On the other side, I personally somehow expect a copying when use any kind of serialization (maybe because the other libs I use do it).
Anyway, thank you for your great work!

**Comment by jtv** [comment]
You make a good case.  Next question: when do I find time?  Probably when the weather gets colder...  :-)

**Comment by jtv** [comment]
Perhaps I was a little pessimistic.  It's not that hard.  I finally figured out why I supported `to_buf()` for `pqxx::zview` but not for `std::string_view` — the reason was that `to_buf(pqxx::zview)` could just return its original view, but we can't do that for `std::string_view` because it's not zero-terminated.  And so it's easy to miss that `to_buf(std::string_view)` needs to copy the entire string, whereas all `to_buf(pqxx::zview)` needs to do is just return the original view.

But yes, it's probably better just to live with that cost.

**Comment by jtv** [comment]
@PetrSilar could you try the branch in #730 to see if that solves your problem?

**Comment by PetrSilar** [comment]
I will and let you know ASAP.
&NewLine;
One other thought: Maybe replace this cycle from `Container`'s `array_string_traits::into_buf()`
```c++
for (char const c : elt_traits::to_buf(end - elt_budget, end, elt))
{
  if (c == '\\' or c == '"')
    *here++ = '\\';
  *here++ = c;
}
```
with a copying of the whole value into `here`. i.e. to use `elt_traits::into_buf()` instead and then optimistically believe there is not much of `\` or `"` and to move the contents forward if there is any (and to place escaping `\` into the gap).

**Comment by PetrSilar** [comment]
> @PetrSilar could you try the branch in #730 to see if that solves your problem?

Works for me! 👍

**Comment by jtv** [comment]
Excellent!  I'll just merge that and close this ticket then.  Thanks for the report, and thanks for persisting.

I'll keep the escaping optimisation in mind for later.  Will need some profiling and statistics I think.

**Comment by PetrSilar** [comment]
> Excellent! I'll just merge that and close this ticket then. Thanks for the report, and thanks for persisting.
> 
You're welcome!
> I'll keep the escaping optimisation in mind for later. Will need some profiling and statistics I think.
> 
Definitely.
